blob/service.go

```diff
@@ -211,84 +211,50 @@ func (s *Service) getByCommitment(
 		}
 		return nil, nil, err
 	}
+
 	for _, row := range namespacedShares {
-		if len(row.Shares) == 0 {
-			break
+		appShares, err := toAppShares(row.Shares...)
+		if err != nil {
+			return nil, nil, err
 		}
-
-		rawShares = append(rawShares, row.Shares...)
+		rawShares = append(rawShares, appShares...)
 		proofs = append(proofs, row.Proof)
 
-		// reconstruct the `blobShare` from the first rawShare in range
-		// in order to get blob's length(first share will contain this info)
-		if blobShare == nil {
-			for i, shr := range rawShares {
-				bShare, err := shares.NewShare(shr)
-				if err != nil {
-					return nil, nil, err
-				}
-
-				// ensure that the first share is not a NamespacePaddingShare
-				// these shares are used to satisfy the non-interactive default rules
-				// and are not the part of the blob, so should be removed.
-				isPadding, err := bShare.IsPadding()
-				if err != nil {
-					return nil, nil, err
-				}
-				if isPadding {
-					continue
-				}
-
-				blobShare = bShare
-				// save the length.
-				length, err := blobShare.SequenceLen()
-				if err != nil {
-					return nil, nil, err
-				}
-				amount = shares.SparseSharesNeeded(length)
-				rawShares = rawShares[i:]
-				break
-			}
-		}
-
-		// move to the next row if the blob is incomplete.
-		if amount > len(rawShares) {
-			continue
-		}
-
-		blob, same, err := constructAndVerifyBlob(rawShares[:amount], commitment)
+		var blobs []*Blob
+		blobs, rawShares, err = buildBlobsIfExist(rawShares)
 		if err != nil {
 			return nil, nil, err
 		}
-		if same {
-			return blob, &proofs, nil
+		for _, b := range blobs {
+			if b.Commitment.Equal(commitment) {
+				return b, &proofs, nil
+			}
+			// Falling under this flag means that the data from the last row
+			// was insufficient to create a complete blob. As a result,
+			// the first blob received spans two rows and includes proofs
+			// for both of these rows. All other blobs in the result will relate
+			// to the current row and have a single proof.
+			if spansMultipleRows {
+				spansMultipleRows = false
+				// leave proof only for the current row
+				proofs = proofs[len(proofs)-1:]
+			}
 		}
 
-		// drop info of the checked blob
-		rawShares = rawShares[amount:]
 		if len(rawShares) > 0 {
-			// save proof for the last row in case we have rawShares
-			proofs = proofs[len(proofs)-1:]
-		} else {
-			// otherwise clear proofs
-			proofs = nil
+			spansMultipleRows = true
+			continue
 		}
-		blobShare = nil
-	}
-
-	if len(rawShares) == 0 {
-		return nil, nil, ErrBlobNotFound
+		proofs = nil
 	}
 
-	blob, same, err := constructAndVerifyBlob(rawShares, commitment)
-	if err != nil {
-		return nil, nil, err
+	err = ErrBlobNotFound
+	if len(rawShares) > 0 {
+		reportErr := fmt.Errorf("THIS IS A BUG: INCOMPLETE BLOB DETECTED at: %d height", height)
```

Comment:
```suggestion
		err = fmt.Errorf("incomplete blob detected at %d: %w", height, err)
```

This might not be a bug necessary, but a malformed block. In this case you can wrap error without join